add doc on functions asynchronous connection (#12966)
kyma-project/kyma: docs/03-tutorials/00-serverless/svls-11-set-asynchronous-connection-of-functions.md

---
title: Set asynchronous communication between Functions
---

This tutorial shows how you can connect two Functions using asynchronous communication. You can do it thanks to the In-cluster Eventing embedded in Kyma. The tutorial is based on Function from [Incluster Eventing example](https://github.com/kyma-project/examples/pull/188).

> 

Reviewer: We should add scenario description:
I.e: 

In this example we will demonstrate how to connect two function asynchronously.

We can think of very simple scenario of two functions, where the first function will be accepting incoming traffic via HTTP, sanitising the payload and publishing the content as an in-cluster event via kyma eventing.
Another function will be message receiver. It will subscribe to the given event type and will store the payload.

## Prerequisites

Before you start, make sure you have these tools installed:

- [Kyma CLI](https://github.com/kyma-project/cli)
- Kyma installed locally or on a cluster

## Steps

1. Export the `KUBECONFIG` variable:
   ```bash
   export KUBECONGIG={KUBECONFIG_PATH}
   ```
2. Create two folders in your project - `emitter` and `receiver`.
3. Go to your `emitter` folder and run the `init` Kyma CLI command to initiate your Function:

   ```bash
   kyma init function
    ```

  The `init` command creates these files in your workspace folder:

  - `config.yaml`	with the Function's configuration

>**NOTE:** See the detailed description of all fields available in the [`config.yaml` file](../../05-technical-reference/svls-06-function-configuration-file.md).

  - `handler.js` with the Function's code and the simple "Hello Serverless" logic
  
  - `package.json` with the Function's dependencies

> 4. In the `config.yaml` file configure your API Rule to allow for the incoming outside traffic (using HTTP). Enter your cluster domain in the host key:

Reviewer: ```suggestion
4. In the `config.yaml` file configure an API Rule to expose your function and allow for incoming traffic over HTTP. Enter your cluster domain in the host key:
```

  ```yaml
  apiRules:
    - name: incomming-http-trigger
      service:
        host: incomming.inclusterevent.acme.shoot.canary.k8s-hana.ondemand.com
      rules:
        - methods:
            - GET
          accessStrategies:
            - handler: allow
  ```

> 5. Provide your Function logic in the `handler.js` file. In the example you can find a simple code sanitization:

Reviewer: ```suggestion
5. Provide your Function logic in the `handler.js` file. 

In the following snippet you will not find any actual sanitization logic - here, `sanitise` function is just a placeholder in this example :
```

   ```js
   const { v4: uuidv4 } = require('uuid');
   module.exports - {
       main: function (event, context) {
           let sanitsedData = sanitise(event.data)
           var eventOut=event.buildResponseCloudEvent(uuidv4(), "sap.kyma.custom.acme.payload.sanitised.v1", sanitisedData);
           eventOut.source="kyma"
           eventOut.specversion="1.0"
           event.publishCloudEvent(eventOut);
           console.log(`Payload pushed to sap.kyma.custom.acme.payload.sanitised.v1`,eventOut)
           return eventOut;
       }
   }
   let sanitise = (data)=>{
       console.log(`sanitising data...`)
       console.log(data)
       return data
   }
   ```

> 

Reviewer: add some extra guidance:

Ie:

Please note the `sap.kyma.custom.acme.payload.sanitised.v1`. This is the event type the emitter is publishing to. The one used here is an  example of-course and you can choose a  different one that better describes your usecase.  Keep in mind the following constraints (pls validate with @k15r ):
 -  it must consist of seven elements separated by dots . 
 -   First three elements must remain unchanged (`sap.kyma.custom`).
 -  The last element specifies the version of the event.

The receiving function must subscribe to this event type. 

Also...

Please note that `event` object provides convinience functions to build and publish events (the content of point 7 bellongs here)

6. Apply your emitter Function:

  ```bash
  kyma apply function
  ```

> Your emitter Function sends the event, altered in a required way, to the receiver Function.

Reviewer: ```suggestion
Having applied it, your function will be built and deployed in kyma runtime. Kyma will expose it via API rule. Any incomming payload would be processed by your emitter Function which will send a sanitised content to whatever workload that subscribes to selected event type - in our case - Receiver function.
```

> 7. To send the event you need to build the Cloud Event. To learn more please visit the [Function's Specification](https://kyma-project.io/docs/kyma/latest/05-technical-reference/svls-08-function-specification#event-object-sdk) page.

Reviewer: 7 th point would better belong as explanation of emitter code - not here IMO

In addition, your `eventOut.source` key needs to point at `“kyma”` to use Kyma In-cluster Eventing.

8. Go to your `receiver` folder and run the `init` Kyma CLI command to initiate your function:
   ```bash
   kyma init function
   ```
9. The `init` command creates the same files as in the emitter folder.
10. In the `config.yaml` file configure your Subscriptions to allow for receipt and storage of the event data:
    ```yaml
    name: event-receiver
    namespace: default
    runtime: nodejs14
    source:
       sourceType: inline
    subscriptions:
       - name: event-receiver
         protocol: ""
         filer:
            filters:
               - eventSource:
                   property: source
                   type: exact
                   value: ""
                eventType:
                   property: type
                   type: exact
                   value: sap.kyma.custom.acme.payload.sanitised.v1
    ```

> 11. The `value` key in your `handler.js` file must consist of seven elements separated by dots, for example `sap.kyma.custom.acme.payload.sanitised.v1`. First three elements must remain unchanged (`sap.kyma.custom`). The last element specifies the version of the event.

Reviewer: the event publisher is in ppower to pick event type. Event receiver must subscribe to whatever emiter chooses as event type.
So its better to explain it as part of the emmiter code ( above)

       return data
   }
   ```
   Please note the `sap.kyma.custom.acme.payload.sanitised.v1`. This is the event type the emitter publishes to. The one used here is an example. You can choose a different one that better suits your use case. Keep in mind the constraints described on the [Event names](https://kyma-project.io/docs/kyma/latest/05-technical-reference/evnt-01-event-names) page.

   Please note that event object provides convinience functions to build and publish events. To send the event you need to build the Cloud Event. To learn more please visit the [Function's Specification](https://kyma-project.io/docs/kyma/latest/05-technical-reference/svls-08-function-specification#event-object-sdk) page. In addition, your `eventOut.source` key needs to point at `“kyma”` to use Kyma In-cluster Eventing.

6. Apply your emitter Function:

    ```bash
    kyma apply function
    ```
   Having applied it, your Function is built and deployed in Kyma runtime. Kyma will expose it via API Rule. Any incoming payload would be processed by your emitter Function. It sends the sanitised content to whatever workload that subscribes to selected event type - in our case - receiver Function.

> 

Reviewer: I think its worth to say here:

Now you can test the first function alone: Send any payload to the first function and see you your HTTP trafic is accepted

```
export KYMA_DOMAIN=.... // export the variable for you kyma domain

curl -X POST https://incomming.${KYMA_DOMAIN}
   -H 'Content-Type: application/json'
   -d '{"foo":"bar"}'
```

> 7. Go to your `receiver` folder and run the `init` Kyma CLI command to initiate your function:

Reviewer: ```suggestion
7. Go to your `receiver` folder and run the `init` Kyma CLI command to initialise the scaffold for your second Function:
```

   ```bash
   kyma init function
   ```
8. The `init` command creates the same files as in the emitter folder.

> 9.  In the `config.yaml` file configure your Subscriptions to allow for receipt and storage of the event data:

Reviewer: ```suggestion
9.  In the `config.yaml` file configure which event types your Function should subscribe to:
```

    ```yaml
    name: event-receiver
    namespace: default
    runtime: nodejs14
    source:
       sourceType: inline
    subscriptions:
       - name: event-receiver
         protocol: ""
         filer:
            filters:
               - eventSource:
                   property: source
                   type: exact
                   value: ""
                eventType:
                   property: type
                   type: exact
                   value: sap.kyma.custom.acme.payload.sanitised.v1
    ```

> 10.  When you call your emitter Function, your receiver Function must also respond.

Reviewer: I would add a step (10) to deploy the second function ( kyma deploy function ).
After you call it function is configured, built and deployed in kyma runtime. The susbscription becomes active and all events with selected tyle will be processed by the function

11. Test the whole setup

Now when you send a payload to the first function ( i.e via above mentioned POST request ) you should be able to observe that thanks to the fact that functions are joined by the in-cluster eventing  the payload visits both of your functions. 
You can see ( i.e in the function logs ) that both sanitization logic ( by the first function)  and the storing logic ( by the second function ) is executed.

Summary:

You can distribute application logic into specialised, event-driven functions

- Kyma installed locally or on a cluster

## Steps

1. Export the `KUBECONFIG` variable:
   ```bash
   export KUBECONGIG={KUBECONFIG_PATH}
   ```
2. Create two folders in your project - `emitter` and `receiver`.
3. Go to your `emitter` folder and run the `init` Kyma CLI command to initialise a scaffold for your first Function:

   ```bash
   kyma init function
    ```

  The `init` command creates these files in your workspace folder:

  - `config.yaml`	with the Function's configuration

>**NOTE:** See the detailed description of all fields available in the [`config.yaml` file](../../05-technical-reference/svls-06-function-configuration-file.md).

  - `handler.js` with the Function's code and the simple "Hello Serverless" logic
  
  - `package.json` with the Function's dependencies

4. In the `config.yaml` file configure an API Rule to expose your Function to the incoming traffic over HTTP. Enter the subdomain name as the `host` property:

    ```yaml
    apiRules:
      - name: incoming-http-trigger
        service:
          host: incoming
        rules:
          - methods:
              - GET
            accessStrategies:
              - handler: allow
    ```

> 5. Provide your Function logic in the `handler.js` file. In the following example you do not find an actual sanitisation logic, `sanitise` Function is just a placeholder:

Reviewer 2: ```suggestion
5. Provide your Function logic in the `handler.js` file. In the following example, there's no actual sanitization logic, `sanitise` Function is just a placeholder:
```

   ```js
   const { v4: uuidv4 } = require('uuid');

>    module.exports - {

Reviewer: ```suggestion
   module.exports = {
```

       main: function (event, context) {
           let sanitsedData = sanitise(event.data)
           var eventOut=event.buildResponseCloudEvent(uuidv4(), "sap.kyma.custom.acme.payload.sanitised.v1", sanitisedData);
           eventOut.source="kyma"
           eventOut.specversion="1.0"
           event.publishCloudEvent(eventOut);
           console.log(`Payload pushed to sap.kyma.custom.acme.payload.sanitised.v1`,eventOut)
           return eventOut;
       }
   }
   let sanitise = (data)=>{
       console.log(`sanitising data...`)
       console.log(data)
       return data
   }
   ```

>    The `sap.kyma.custom.acme.payload.sanitised.v1` is an exemplary event type declared by the emitter Function when publishing events. You can choose a different one that better suits your use case. Keep in mind the constraints described on the [Event names](https://kyma-project.io/docs/kyma/latest/05-technical-reference/evnt-01-event-names) page. The receiver subscribes to the event type to consume the events.

Reviewer 2: ```suggestion
   >**NOTE:** The `sap.kyma.custom.acme.payload.sanitised.v1` is a sample event type declared by the emitter Function when publishing events. You can choose a different one that better suits your use case. Keep in mind the constraints described on the [Event names](https://kyma-project.io/docs/kyma/latest/05-technical-reference/evnt-01-event-names) page. The receiver subscribes to the event type to consume the events.
```

>    Please note that event object provides convinience functions to build and publish events. To send the event you need to build the Cloud Event. To learn more please visit the [Function's Specification](https://kyma-project.io/docs/kyma/latest/05-technical-reference/svls-08-function-specification#event-object-sdk) page. In addition, your `eventOut.source` key needs to point at `“kyma”` to use Kyma In-cluster Eventing.

Reviewer 2: ```suggestion
   >**NOTE:** The event object provides convenience functions to build and publish events. To send the event, build the Cloud Event. To learn more, read [Function's specification](https://kyma-project.io/docs/kyma/latest/05-technical-reference/svls-08-function-specification#event-object-sdk). In addition, your **eventOut.source** key must point to `“kyma”` to use Kyma in-cluster Eventing.
```

6. Apply your emitter Function:

    ```bash
    kyma apply function
    ```

>    Having applied it, your Function is built and deployed in Kyma runtime. Kyma will expose it via API Rule. Any incoming payload would be processed by your emitter Function. It sends the sanitised content to whatever workload that subscribes to selected event type - in our case - receiver Function.

Reviewer 2: ```suggestion
   Your Function is now built and deployed in Kyma runtime. Kyma will expose it via API Rule. Any incoming payload will be processed by your emitter Function. It sends the sanitized content to whatever workload that subscribes to the selected event type - in our case - the receiver Function.
```

7. Test the first Function. Send a paylod and see if your HTTP traffic is accepted:

      ```bash
      export KYMA_DOMAIN=.... # export the variable for you Kyma domain
      
      curl -X POST https://incoming.${KYMA_DOMAIN}
      -H 'Content-Type: application/json'
      -d '{"foo":"bar"}'
      ```

> 

Reviewer 2: Consider adding subheadings that will break the wall of text and make the whole tutorial more user-friendly. For example: 

```
### Create the emitter Function

steps

### Create the receiver Function

steps

````

8. Go to your `receiver` folder and run the `init` Kyma CLI command to to initialise the scaffold for your second Function:
   ```bash
   kyma init function
   ```
9.  The `init` command creates the same files as in the `emitter` folder.
10. In the `config.yaml` file configure which event types your Function should subscribe to:
    ```yaml
    name: event-receiver
    namespace: default
    runtime: nodejs14
    source:
       sourceType: inline
    subscriptions:
       - name: event-receiver
         protocol: ""
         filer:
            filters:
               - eventSource:
                   property: source
                   type: exact
                   value: ""
                eventType:
                   property: type
                   type: exact
                   value: sap.kyma.custom.acme.payload.sanitised.v1
    ```
11.  Apply your receiver Function:
     ```bash
     kyma apply function
     ```
   The Function is configured, built and deployed in Kyma runtime. The Susbscription becomes active and all events with selected type will be processed by the Function.  

12.  Test the whole setup  

> Send a payload to the first Function (for example using the above-mentioned POST request). As the Functions are joined by the In-cluster Eventing, the payload is directed to both of your Functions.

Reviewer: ```suggestion
Send a payload to the first Function (for example using the above-mentioned POST request). As the Functions are joined by the In-cluster Eventing, the payload is processed in sequence by both of your Functions.
```

You can see (for example in the Function logs) that both, sanitization logic (using the first Function) and the storing logic (using the second Function) are executed.

> Summary:

Reviewer: ```suggestion
13. Summary
```

> You can distribute application logic into specialised, event-driven Functions

Reviewer: ```suggestion
This tutorial demonstrated only one possible use case example. You will find plenty of use cases yourself for how to  orchestrate your  application logic into specialised Functions and gain from  decoupled, re-usable components and  event-driven architecture.
```

---
title: Set asynchronous communication between Functions
---

This tutorial demonstrates how to connect two Functions asynchronously. It is based on the [in-cluster Eventing example](https://github.com/kyma-project/examples/tree/main/incluster_eventing).

> The example provides a very simple scenario of asynchronous communication between two Functions. The first Function accepts the incoming traffic via HTTP, sanitizes the payload, and publishes the content as an in-cluster event using [Kyma Eventing](https://kyma-project.io/docs/kyma/latest/01-overview/main-areas/eventing/).

Reviewer 2: ```suggestion
The example provides a very simple scenario of asynchronous communication between two Functions. The first Function accepts the incoming traffic via HTTP, sanitizes the payload, and publishes the content as an in-cluster event using [Kyma Eventing](../../01-overview/main-areas/eventing).
```

The second Function is a message receiver. It subscribes to the given event type and stores the payload.

> It shows only one possible use case. You can find plenty of use cases yourself on how to orchestrate your application logic into specialised Functions and benefit from decoupled, re-usable components and event-driven architecture.

Reviewer 2: ```suggestion
This tutorial shows only one possible use case. There are many more use cases on how to orchestrate your application logic into specialized Functions and benefit from decoupled, re-usable components and event-driven architecture.
```

Reviewer 2: I rephrased it a bit, because ` You can find plenty of use cases yourself...` brings up a question `where can I find it?` and I assume we just want to point out that there are other use cases, not provide the users with them.

## Prerequisites

- [Kyma CLI](https://github.com/kyma-project/cli)
- Kyma installed locally or on a cluster

## Steps

1. Export the `KUBECONFIG` variable:
   ```bash
   export KUBECONFIG={KUBECONFIG_PATH}
   ```

> 2. Create two folders in your project - `emitter` and `receiver`.

Reviewer 2: ```suggestion
2. Create the `emitter` and `receiver` folders in your project.
```

### Create the emitter Function

> 1. Go to the `emitter` folder and run Kyma CLI `init` command to initialize a scaffold for your first Function:

Reviewer 2: ```suggestion
1. Go to the `emitter` folder and run Kyma CLI `init` command to initialize the scaffold for your first Function:
```

    ```yaml
    apiRules:
      - name: incoming-http-trigger
        service:
          host: incoming
        rules:
          - methods:
              - GET
            accessStrategies:
              - handler: allow
    ```

3. Provide your Function logic in the `handler.js` file:
>**NOTE:** In this example, there's no sanitization logic. The `sanitize` Function is just a placeholder.

   ```js
   const { v4: uuidv4 } = require('uuid');
   module.exports = {
       main: function (event, context) {
           let sanitsedData = sanitise(event.data)
           var eventOut=event.buildResponseCloudEvent(uuidv4(), "sap.kyma.custom.acme.payload.sanitised.v1", sanitisedData);
           eventOut.source="kyma"
           eventOut.specversion="1.0"
           event.publishCloudEvent(eventOut);
           console.log(`Payload pushed to sap.kyma.custom.acme.payload.sanitised.v1`,eventOut)
           return eventOut;
       }
   }
   let sanitise = (data)=>{
       console.log(`sanitising data...`)
       console.log(data)
       return data
   }
   ```

>    >**NOTE:** The `sap.kyma.custom.acme.payload.sanitised.v1` is a sample event type declared by the emitter Function when publishing events. You can choose a different one that better suits your use case. Keep in mind the constraints described on the [Event names](https://kyma-project.io/docs/kyma/latest/05-technical-reference/evnt-01-event-names) page. The receiver subscribes to the event type to consume the events.

Reviewer 2: ```suggestion
   >**NOTE:** The `sap.kyma.custom.acme.payload.sanitised.v1` is a sample event type declared by the emitter Function when publishing events. You can choose a different one that better suits your use case. Keep in mind the constraints described on the [Event names](../../05-technical-reference/evnt-01-event-names.md) page. The receiver subscribes to the event type to consume the events.
```

Reviewer 2: Use relative links to link to the docs within the Kyma project. Otherwise, they will appear as external links on the website:

<img width="745" alt="Screen Shot 2022-01-21 at 11 19 42" src="https://user-images.githubusercontent.com/35192450/150510449-19013b0d-0917-4696-a855-08565b9b6c85.png">

>    >**NOTE:** The event object provides convenience functions to build and publish events. To send the event, build the Cloud Event. To learn more, read [Function's specification](https://kyma-project.io/docs/kyma/latest/05-technical-reference/svls-08-function-specification#event-object-sdk). In addition, your **eventOut.source** key must point to `“kyma”` to use Kyma in-cluster Eventing.

Reviewer 2: ```suggestion
   >**NOTE:** The event object provides convenience functions to build and publish events. To send the event, build the Cloud Event. To learn more, read [Function's specification](../../05-technical-reference/svls-08-function-specification#event-object-sdk). In addition, your **eventOut.source** key must point to `“kyma”` to use Kyma in-cluster Eventing.
```

4. Apply your emitter Function:

    ```bash
    kyma apply function
    ```
   Your Function is now built and deployed in Kyma runtime. Kyma exposes it through the API Rule. The incoming payloads are processed by your emitter Function. It then sends the sanitized content to the workload that subscribes to the selected event type. In our case, it's the receiver Function.

5. Test the first Function. Send the payload and see if your HTTP traffic is accepted:

      ```bash
      export KYMA_DOMAIN={KYMA_DOMAIN_VARIABLE}
      
      curl -X POST https://incoming.${KYMA_DOMAIN}
      -H 'Content-Type: application/json'
      -d '{"foo":"bar"}'
      ```
### Create the receiver Function

> 1. Go to your `receiver` folder and run Kyma CLI `init` Kyma CLI command to initialize the scaffold for your second Function:

Reviewer 2: ```suggestion
1. Go to your `receiver` folder and run Kyma CLI `init` command to initialize the scaffold for your second Function:
```

   ```bash
   kyma init function
   ```

> 2.  The `init` command creates the same files as in the `emitter` folder.

Reviewer 2: We can make it a regular sentence under the command, not necessarily another step 🤔 Especially that you made it a regular sentence in the previous paragraph.

3. In the `config.yaml` file, configure event types your Function will subscribe to:
    ```yaml
    name: event-receiver
    namespace: default
    runtime: nodejs14
    source:
       sourceType: inline
    subscriptions:
       - name: event-receiver
         protocol: ""
         filer:
            filters:
               - eventSource:
                   property: source
                   type: exact
                   value: ""
                eventType:
                   property: type
                   type: exact
                   value: sap.kyma.custom.acme.payload.sanitised.v1
    ```
4.  Apply your receiver Function:
     ```bash
     kyma apply function
     ```
   The Function is configured, built, and deployed in Kyma runtime. The Subscription becomes active and all events with the selected type are processed by the Function.  

> 5.  Test the whole setup  

Reviewer 2: ```suggestion
### Test the whole setup  
```

Reviewer 2: I assume this is not part of the `Create the receiver Function` paragraph, rather a new one?

> Send a payload to the first Function (for example using the above-mentioned POST request). As the Functions are joined by the In-cluster Eventing, the payload is processed in sequence by both of your Functions.

Reviewer 2: ```suggestion
Send a payload to the first Function (for example using the above-mentioned POST request). As the Functions are joined by the In-cluster Eventing, the payload is processed in sequence by both of your Functions.
```
```suggestion
Send a payload to the first Function. For example, use the POST request mentioned above. As the Functions are joined by the in-cluster Eventing, the payload is processed in sequence by both of your Functions.
```

> You can see (for example in the Function logs) that both, sanitization logic (using the first Function) and the storing logic (using the second Function) are executed.

Reviewer 2: ```suggestion
In the Function's logs, you can see that both sanitization logic (using the first Function) and the storing logic (using the second Function) are executed.
```

## Initial steps

1. Export the `KUBECONFIG` variable:
   ```bash
   export KUBECONFIG={KUBECONFIG_PATH}
   ```
2. Create two folders in your project - `emitter` and `receiver`.

## Create the emitter Function

1. Go to the `emitter` folder and run Kyma CLI `init` command to initialize a scaffold for your first Function:

   ```bash
   kyma init function
    ```

  The `init` command creates these files in your workspace folder:

  - `config.yaml`	with the Function's configuration

   >**NOTE:** See the detailed description of all fields available in the [`config.yaml` file](../../05-technical-reference/svls-06-function-configuration-file.md).

  - `handler.js` with the Function's code and the simple "Hello Serverless" logic
  
  - `package.json` with the Function's dependencies

2. In the `config.yaml` file, configure an API Rule to expose your Function to the incoming traffic over HTTP. Provide the subdomain name in the `host` property:

    ```yaml
    apiRules:
      - name: incoming-http-trigger
        service:
          host: incoming
        rules:
          - methods:
              - GET
            accessStrategies:
              - handler: allow
    ```

> 3. Provide your Function logic in the `handler.js` file. In the following example, there's no actual sanitization logic, `sanitise` Function is just a placeholder:

Reviewer 2: ```suggestion
3. Provide your Function logic in the `handler.js` file:
>**NOTE:** In this example, there's no sanitization logic. The `sanitize` Function is just a placeholder.
```

>**NOTE:** In this example, there's no sanitization logic. The `sanitize` Function is just a placeholder.

   ```js
   const { v4: uuidv4 } = require('uuid');
   module.exports = {
       main: function (event, context) {
           let sanitsedData = sanitise(event.data)
           var eventOut=event.buildResponseCloudEvent(uuidv4(), "sap.kyma.custom.acme.payload.sanitised.v1", sanitisedData);
           eventOut.source="kyma"
           eventOut.specversion="1.0"
           event.publishCloudEvent(eventOut);
           console.log(`Payload pushed to sap.kyma.custom.acme.payload.sanitised.v1`,eventOut)
           return eventOut;
       }
   }
   let sanitise = (data)=>{
       console.log(`sanitising data...`)
       console.log(data)
       return data
   }
   ```
   >**NOTE:** The `sap.kyma.custom.acme.payload.sanitised.v1` is a sample event type declared by the emitter Function when publishing events. You can choose a different one that better suits your use case. Keep in mind the constraints described on the [Event names](../../05-technical-reference/evnt-01-event-names.md) page. The receiver subscribes to the event type to consume the events.

>    >**NOTE:** The event object provides convenience functions to build and publish events. To send the event, build the Cloud Event. To learn more, read [Function's specification](../../05-technical-reference/svls-08-function-specification#event-object-sdk). In addition, your **eventOut.source** key must point to `“kyma”` to use Kyma in-cluster Eventing.

Reviewer 2: ```suggestion
   >**NOTE:** The event object provides convenience functions to build and publish events. To send the event, build the Cloud Event. To learn more, read [Function's specification](../../05-technical-reference/svls-08-function-specification.md#event-object-sdk). In addition, your **eventOut.source** key must point to `“kyma”` to use Kyma in-cluster Eventing.
```

    ```bash
    kyma apply function
    ```

>    Your Function is now built and deployed in Kyma runtime. Kyma will expose it via API Rule. Any incoming payload will be processed by your emitter Function. It sends the sanitized content to whatever workload that subscribes to the selected event type - in our case - the receiver Function.

Reviewer 2: ```suggestion
   Your Function is now built and deployed in Kyma runtime. Kyma exposes it through the API Rule. The incoming payloads are processed by your emitter Function. It then sends the sanitized content to the workload that subscribes to the selected event type. In our case, it's the receiver Function.
```

> 5. Test the first Function. Send a paylod and see if your HTTP traffic is accepted:

Reviewer 2: ```suggestion
5. Test the first Function. Send the payload and see if your HTTP traffic is accepted:
```

      ```bash

>       export KYMA_DOMAIN=.... # export the variable for you Kyma domain

Reviewer 2: ```suggestion
      export KYMA_DOMAIN={KYMA_DOMAIN_VARIABLE}
```

      
      curl -X POST https://incoming.${KYMA_DOMAIN}
      -H 'Content-Type: application/json'
      -d '{"foo":"bar"}'
      ```
## Create the receiver Function

> 1. Go to your `receiver` folder and run the `init` Kyma CLI command to to initialise the scaffold for your second Function:

Reviewer 2: ```suggestion
1. Go to your `receiver` folder and run Kyma CLI `init` Kyma CLI command to initialize the scaffold for your second Function:
```

   ```bash
   kyma init function
   ```
2.  The `init` command creates the same files as in the `emitter` folder.

> 3. In the `config.yaml` file configure which event types your Function should subscribe to:

Reviewer 2: ```suggestion
3. In the `config.yaml` file, configure event types your Function will subscribe to:
```

    ```yaml
    name: event-receiver
    namespace: default
    runtime: nodejs14
    source:
       sourceType: inline
    subscriptions:
       - name: event-receiver
         protocol: ""
         filer:
            filters:
               - eventSource:
                   property: source
                   type: exact
                   value: ""
                eventType:
                   property: type
                   type: exact
                   value: sap.kyma.custom.acme.payload.sanitised.v1
    ```
4.  Apply your receiver Function:
     ```bash
     kyma apply function
     ```

>    The Function is configured, built and deployed in Kyma runtime. The Susbscription becomes active and all events with selected type will be processed by the Function.  

Reviewer 2: ```suggestion
   The Function is configured, built, and deployed in Kyma runtime. The Subscription becomes active and all events with the selected type are processed by the Function.  
```

> 5.  Test the whole setup  

Reviewer 2: ```suggestion
## Test the whole setup  
```

> Send a payload to the first Function (for example using the above-mentioned POST request). As the Functions are joined by the In-cluster Eventing, the payload is processed in sequence by both of your Functions.

Reviewer 2: ```suggestion
Send a payload to the first Function. For example, use the POST request mentioned above. As the Functions are joined by the in-cluster Eventing, the payload is processed in sequence by both of your Functions.
```

> You can see (for example in the Function logs) that both, sanitization logic (using the first Function) and the storing logic (using the second Function) are executed.

Reviewer 2: ```suggestion
In the Function's logs, you can see that both sanitization logic (using the first Function) and the storing logic (using the second Function) are executed.
```

> ## Summary

Reviewer 2: I'd add it as a note in the overview.

> This tutorial demonstrated only one possible use case. You can find plenty of use cases yourself on how to orchestrate your application logic into specialised Functions and benefit from decoupled, re-usable components and event-driven architecture.

Reviewer 2: ```suggestion
This tutorial demonstrated only one possible use case. There are many more use cases on how to orchestrate your application logic into specialized Functions and benefit from decoupled, re-usable components and event-driven architecture.
```
